### PR TITLE
Allow setting link_state in spec conf

### PR DESCRIPTION
**Pull request by booxter**
When link_state is set, the link state of the corresponding VF will be
set to the passed value. Allowed values: auto, enable, disable. This is
useful in testing when e.g. in-PF switching should be enabled for
allocated VFs (which allows to connect pods running on the same host
without external switch or PF being plugged in the network).

### CODE REVIEW

**Comment by booxter** [comment]
@ahalim-intel @zshi-redhat mellanox tests failed but it doesn't immediately seem related. How can I retrigger the suite?

**Comment by ahalim-intel** [comment]
cc @mellanox-ci

**Comment by booxter** [comment]
@ahalim-intel @zshi-redhat now that netlink merged their patch to support vf link state updates, this is now ready to review / merge. Thanks!

File: pkg/config/config.go [hunk 1]
```diff
@@ -70,6 +70,11 @@ func LoadConf(bytes []byte) (*sriovtypes.NetConf, error) {
 		return nil, fmt.Errorf("LoadConf(): vlan QoS PCP %d invalid: value must be in the range 0-7", n.VlanQoS)
 	}
 
+	// validate that vlan id is set if vlan qos is configured
```
Comment (booxter) [inline pkg/config/config.go]: this comment should be fixed...

File: pkg/sriov/sriov.go [hunk 2]
```diff
@@ -95,6 +96,12 @@ func (n *MyNetlink) LinkSetVfTrust(link netlink.Link, vf int, state bool) error
 	return netlink.LinkSetVfTrust(link, vf, state)
 }
 
+// LinkSetVfState using NetlinkManager
+func (n *MyNetlink) LinkSetVfState(link netlink.Link, vf int, state uint32) error {
+	return netlink.LinkSetVfState(link, vf, state)
+}
+
+// LinkSetVfState using NetlinkManager
```
Comment (booxter) [inline pkg/sriov/sriov.go]: this comment should be removed...

**Comment by booxter** [comment]
@killianmuldoon (considering that @ahalim-intel added you as a reviewer) do you think you could find time to review this patch in the near future? Thanks!

@ahalim-intel do we need other reviews in addition to @killianmuldoon's?

**Comment by ahalim-intel** [comment]
Hi @booxter, @killianmuldoon is on leave atm. I will review it myself, basically some smoke tests. Code changes look fine. I very much welcome if anyone else wants to review and test this and give some feedback.

The Mellanox CI is failing some other reason.

**Comment by killianmuldoon** [comment]
@booxter I'm running into an issue getting this to work with Multus  ( IPAM plugin returned missing IP config ). I'm going to see if I can track down whether the origin of this issue is my config or something in the CNI.

**Comment by booxter** [comment]
@killianmuldoon let me know more about the error, maybe I can help. In general, the latest CNI plugin should be able to work even with empty `ipam` section in net-attach-def CRD.

**Comment by killianmuldoon** [comment]
@booxter it seems to work if I remove the IPAM configuration from the network CRD  - but I'm running into a consistent error when I try to schedule it with the IPAM on. The output from Multus "is error in invoke Delegate add - "sriov": IPAM plugin returned missing IP config." The pod then sticks in container creating as it can't set up the network.

I'm still trying to track down the root of the error.

**Comment by killianmuldoon** [comment]
@booxter You can ignore this issue now - I tracked this down to a Multus configuration issue on my end- but thanks for the help! .

**Comment by killianmuldoon** [comment]
@booxter Can I ask what is the expected behaviour if the link-state is set to "auto" ? Is this supposed to set the link dynamically as the PF state changes ? 

Also, if the VF is set to down in the kernel and it's set to "auto" in the SRIOV config what's the expected state of the interface in the container?

**Comment by booxter** [comment]
@killianmuldoon `auto` means reflect PF state to its VFs (which is the default behavior when you don't set anything). I left it supported as an explicit value just for completeness with what `ip link` supports.

`auto` will reflect PF state; `disabled` will keep VF always `DOWN` regardless of whether PF link is `UP`; `enabled` is to keep VF `UP` regardless of whether PF link is `DOWN`. This is exactly the same as what `ip link` man page describes.

**Comment by booxter** [comment]
https://www.systutorials.com/docs/linux/man/8-ip-link/

```
 state auto|enable|disable - set the virtual link state as seen by the specified VF. Setting to auto means a reflection of the PF link state, enable lets the VF to communicate with other VFs on this host even if the PF link state is down, disable causes the HW to drop any packets sent by the VF. 
```

**Comment by killianmuldoon** [comment]
@booxter On my set up this isn't working where the PF is down - i.e. not physically connected. In that case the ip link cli returns an "Operation not permitted" error from Netlink. Can you share details of a set up where this is working?

**Comment by booxter** [comment]
@killianmuldoon some older drivers / NICs don't support it, in which case you get the error. This is probably what happens to you. I used `i40e` and it works. But `igb` does not.

**Comment by killianmuldoon** [comment]
@booxter  exactly right - I was using it on an igb  device.  I can see it working now (disable, enable, auto) on a device that has no physical connection. Do we have some sort of list of which devices support this functionality? It could be helpful to link to some documentation covering the issues I ran into here in the readme.

**Comment by booxter** [comment]
@killianmuldoon I don't have a list (I was confused by the fact it doesn't work everywhere too), but we could perhaps add a general note in docs saying it may require driver support, and list these two as examples of good and bad drivers. Will it be good enough?

**Comment by killianmuldoon** [comment]
@booxter That sounds good to me - at least we can flag that there are some issues there - I struggled to find good information about the link-state setting through netlink overall so I'm not surprised driver compatibility isn't easily available.

**Comment by booxter** [comment]
@killianmuldoon I've updated the README document to indicate the feature depends on driver support. Let me know if it's good enough to merge now, or we need something else here. Thanks for review!

File: pkg/sriov/sriov.go [hunk 3]
```diff
@@ -382,5 +407,10 @@ func (s *sriovManager) ResetVFConfig(conf *sriovtypes.NetConf) error {
 		return fmt.Errorf("failed to disable rate limiting for vf %d %v", conf.VFID, err)
 	}
 
+	// Reset link state to `auto`
```
Comment (zshi-redhat) [inline pkg/sriov/sriov.go]: Do we need `if` condition here to make sure `conf.LinkState` is not empty?

Comment (booxter) [reply pkg/sriov/sriov.go]: Not sure. I assume that `auto` is always the right default state to leave a VF in. Unconditional application will also revert to `auto` even when e.g. `NetworkAttachmentDefinition` is updated after the port is initialized but before the port is deleted, which seems like the right thing to do.

It's easy to add a conditional here, let me know if you still believe it's the right thing to do, and I will update the PR.

Thanks!

Comment (booxter) [reply pkg/sriov/sriov.go]: @zshi-redhat please check above

Comment (zshi-redhat) [reply pkg/sriov/sriov.go]: @booxter, one thing I'm not sure is that if a device doesn't support setting link state(whether it's auto, enable or disable), will it fail or ignore the netlink setting?

Comment (booxter) [reply pkg/sriov/sriov.go]: @zshi-redhat oh that's a good point! maybe we can ignore the error there. Or introduce a conditional as you originally proposed and make it less magick-y. Let me respin the PR.

Comment (booxter) [reply pkg/sriov/sriov.go]: @zshi-redhat let me know if the latest version satisfies your request.

**Comment by zshi-redhat** [comment]
looks like Mellanox CI failed due to `no IP addresses available in range set`, @mellanox-ci can we re-run the job and see if it gives green?

**Comment by zshi-redhat** [comment]
Tested this with a no-carrier PF device, all the link states can be configured and restored correctly by SR-IOV CNI.

One observation was:
when `link_state` set to `enable`, `trust` has to be set to `on` in order for VFs from different pods to ping each other. Is this expected?

**Comment by booxter** [comment]
@zshi-redhat I don't think I had to touch `trust` to make it work on `i40e` (at least here on my machine). Is it maybe the default value for `trusted` to have it `on`? Then it could explain that I didn't need to touch it to make local switching work.

**Comment by zshi-redhat** [comment]
> @zshi-redhat I don't think I had to touch `trust` to make it work on `i40e` (at least here on my machine). Is it maybe the default value for `trusted` to have it `on`? Then it could explain that I didn't need to touch it to make local switching work.

The default value of `trust` on my server is `off`, I'm using CentOS 7.6.
did you configure MAC address when enabling link auto?

**Comment by booxter** [comment]
@zshi-redhat I think you may be onto something: default `kubevirt` deployment has `kubemacpool` deployed, which will set a MAC address to each pod / VMI created. Is `trusted` related to custom MAC addresses?

**Comment by zshi-redhat** [comment]
> @zshi-redhat I think you may be onto something: default `kubevirt` deployment has `kubemacpool` deployed, which will set a MAC address to each pod / VMI created. Is `trusted` related to custom MAC addresses?

I didn't have MAC configured when testing `link state`,  guess that's the reason it failed;
re-test with MAC, it works. I think it's because without specified MAC, adminstrative mac is not equal effective mac, and when `trust` being off, `ping` with effective mac as source mac is not allowed.

**Comment by booxter** [comment]
@killianmuldoon let me know if I addressed your concern.

@ahalim-intel let me know how I can retrigger mellanox CI. Or should we just ignore mellanox failure?

**Pull request merged**